### PR TITLE
[#159] script-generator

### PR DESCRIPTION
**Pull request by jmfayard**
# What ?

Generate a Kotlin script from an existing YAML workflow

```bash
export URL=https://raw.githubusercontent.com/jmfayard/refreshVersions/main/.github/workflows/publish-mkdocs-website.yml
./gradlew :script-generator:run --args $URL
```

See `script-generator/README.md`

# Why ?

When lots of suffering has been put into making a YAML workflow to work, people are reluctant to start over from scratch with a library they have yet to know. Better the devil you know... And we don't really plan to change the YAML workflow a lot anyway, what's the point ?

Having a script-generator that gives you quickly a script that generates some YAML that you can compare with your existing workflow via https://www.yamldiff.com/ can be a game changer

# How ?

the complete pull request is huge, but the most interesting parts are regrouopes in small commits

kotlinx-serialization with kaml, plus a pre-processing step because GitHub does crazy shit with its YAML workflows.

kotlinpoet code, with a handy. `Collection.joinToCode()` utility

I try to re-use when possible code that was made for generating the YAML inside the `library` module, using then the new annotation `@InternalGithubActionsApi`

The integration tests from the folder `yaml-input` cover a lots, in my experience the smallest mistake is detected soon with them. And we can add new integration tests easily if we have interesting existing YAML workflows

### CODE REVIEW

File: script-generator/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+dependencies {
+    implementation(project(":wrapper-generator"))
+    implementation(project(":library"))
+    implementation("com.charleskorn.kaml:kaml:0.42.0")
+    implementation("com.squareup:kotlinpoet:1.10.2")
+
+    testImplementation(project(":library"))
+    testImplementation("io.kotest:kotest-assertions-core:5.1.0")
+    testImplementation("io.kotest:kotest-runner-junit5:5.1.0")
+    testImplementation(project(":library"))
```
Comment (krzema12) [inline script-generator/build.gradle.kts:25]: Duplicated (see L22), and also if it's given in `implementation`, it's not needed to specify it in `testImplementation`

**Review by krzema12 — CHANGES_REQUESTED** [review]
I reviewed only the API, the docs, and changes in the library and wrapper-generator. I leave the implementation to you, as agreed about the ownership :)

File: script-generator/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    useJUnitPlatform()
+}
+
+application {
+    mainClass.set("it.krzeminski.githubactions.scriptgenerator.MainKt")
+    tasks.run.get().workingDir = rootProject.projectDir
+}
+
+tasks.getByName("run") {
+    finalizedBy(":library:ktlintFormat")
```
Comment (krzema12) [inline script-generator/build.gradle.kts:38]: Should it say `:library`? I guess it's a leftover after copying from wrapper-generator.

File: script-generator/src/main/kotlin/it/krzeminski/githubactions/scriptgenerator/Jobs.kt [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+import it.krzeminski.githubactions.wrappergenerator.domain.WrapperRequest
+import it.krzeminski.githubactions.wrappergenerator.generation.buildActionClassName
+import it.krzeminski.githubactions.wrappergenerator.wrappersToGenerate
+
+fun YamlWorkflow.generateJobs() = CodeBlock { builder ->
+    jobs.forEach { (name, job) ->
+        builder.add(
+            "job(%S, %M",
+            name,
+            enumMemberName<RunnerType>(job.runsOn) ?: enumMemberName(RunnerType.UbuntuLatest)
```
Comment (krzema12) [inline script-generator/src/main/kotlin/it/krzeminski/githubactions/scriptgenerator/Jobs.kt:18]: I'm trying to convert https://raw.githubusercontent.com/krzema12/vis-assert/master/.github/workflows/deploy.yml but the tool returns

```
WARNING: Unexpected enum class it.krzeminski.githubactions.domain.RunnerType macos-latest=macos-latest expected=[Lit.krzeminski.githubactions.domain.RunnerType;@5a12c728
```

Looks like a bug, and I'm wondering if it should just produce a warning, or make the whole run fail (maybe when adding some flag like `--force`, it should behave like now?).

**Comment by krzema12** [comment]
I added the [`CODEOWNERS file`](https://github.com/krzema12/github-actions-kotlin-dsl/blob/main/.github/CODEOWNERS), please add this line there:

```
/script-generator/ @jmfayard
```

File: script-generator/src/test/kotlin/test/CaseEnumSerializerTest.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.matchers.shouldBe
+import it.krzeminski.githubactions.domain.triggers.PullRequest
+import it.krzeminski.githubactions.scriptmodel.YamlWorkflowTriggers
+import it.krzeminski.githubactions.scriptmodel.myYaml
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.encodeToString
+
+class CaseEnumSerializerTest : FunSpec({
```
[the remark this answers is not in the document]
Comment (jmfayard) [reply script-generator/src/test/kotlin/test/CaseEnumSerializerTest.kt:11]: Done

File: script-generator/README.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
+## 🎉 Generate your Kotlin Script from your existing YAML workflow
```
[the remark this answers is not in the document]
Comment (jmfayard) [reply script-generator/README.md:1]: Done, would appreciate if you can review  & improve the mkdocs page that I created

File: script-generator/build.gradle.kts (hunk 2 above)
Comment (jmfayard) [reply script-generator/build.gradle.kts:38]: Good catch, fixed

File: script-generator/build.gradle.kts (hunk 1 above)
Comment (jmfayard) [reply script-generator/build.gradle.kts:25]: Fixed

**Comment by jmfayard** [comment]
@krzema12 Can you have a look at the doc? https://github.com/krzema12/github-actions-kotlin-dsl/blob/7f4724343b47f4f02c6935ea63b802a585dd6306/docs/user-guide/script-generator.md

I'm done with the rest.

**Review by krzema12 — APPROVED** [review]

**Comment by krzema12** [comment]
Thank you for this awesome piece of work :) Kudos!

**Comment by jmfayard** [comment]
thanks a lot :)

**Pull request merged**